repose-aggregator/components/filters/saml-policy-translation-filter/src/main/scala/org/openrepose/filters/samlpolicy/SamlPolicyTranslationFilter.scala

```diff
@@ -21,21 +21,184 @@
+import javax.servlet.{FilterChain, ServletInputStream, ServletRequest, ServletResponse}
+import javax.ws.rs.core.MediaType
 
+import com.google.common.cache.{Cache, LoadingCache}
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import net.sf.saxon.s9api.XsltExecutable
+import org.openrepose.commons.utils.http.CommonHttpHeader.CONTENT_TYPE
+import org.openrepose.commons.utils.servlet.http.{HttpServletRequestWrapper, HttpServletResponseWrapper, ResponseMode}
 import org.openrepose.core.filter.AbstractConfiguredFilter
 import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.core.services.serviceclient.akka.AkkaServiceClientFactory
 import org.openrepose.filters.samlpolicy.config.SamlPolicyConfig
+import org.openrepose.nodeservice.atomfeed.{AtomFeedListener, AtomFeedService, LifecycleEvents}
+import org.w3c.dom.Document
 
 /**
   * Created by adrian on 12/12/16.
   */
 @Named
-class SamlPolicyTranslationFilter @Inject()(configurationService: ConfigurationService)
-  extends AbstractConfiguredFilter[SamlPolicyConfig](configurationService){
+class SamlPolicyTranslationFilter @Inject()(configurationService: ConfigurationService,
+                                            atomFeedService: AtomFeedService,
+                                            akkaServiceClientFactory: AkkaServiceClientFactory)
+  extends AbstractConfiguredFilter[SamlPolicyConfig](configurationService)
+  with LazyLogging
+  with AtomFeedListener {
 
   override val DEFAULT_CONFIG: String = "saml-policy.cfg.xml"
   override val SCHEMA_LOCATION: String = "/META-INF/config/schema/saml-policy.xsd"
 
-  override def doWork(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit = chain.doFilter(request, response)
+  private var cache: LoadingCache[String, XsltExecutable] = _
+
+  override def doWork(servletRequest: ServletRequest, servletResponse: ServletResponse, chain: FilterChain): Unit = {
+    try {
+      val request = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
+      var response = servletResponse.asInstanceOf[HttpServletResponse]
+      val samlResponse = decodeResponse(request)
+      val rawDocument = readToDom(samlResponse)
+
+      val version = determineVersion(rawDocument)
+      val finalDocument = version match {
+        case 1 =>
+          request.addHeader("Identity-API-Version", "1.0")
+          rawDocument
+        case 2 =>
+          response = new HttpServletResponseWrapper(response, ResponseMode.PASSTHROUGH, ResponseMode.MUTABLE)
+          request.addHeader("Identity-API-Version", "2.0")
+          val issuer = validateResponseAndGetIssuer(rawDocument)
+          val translatedDocument = translateResponse(rawDocument, cache.get(issuer))
+          signResponse(translatedDocument)
+      }
+      val inputStream = convertDocumentToStream(finalDocument)
+
+      request.replaceHeader(CONTENT_TYPE.toString, MediaType.APPLICATION_XML)
```

Comment:
Should the `Content-Length` header also be replaced? Decoding the response will likely change that value, and if I remember correctly, a request requires either the `Content-Length` header or `chunked` encoding.

Not sure how easy that will be since you're dealing in a DOM then a stream.